src/sagas/projects.js

```diff
@@ -112,6 +114,23 @@ export function* projectExported({payload: {exportType}}) {
   }
 }
 
+export function* archiveProject({payload: {projectKey}}) {
+  const state = yield select();
+  const currentProjectKey = getCurrentProjectKey(state);
+  if (projectKey === currentProjectKey) {
+    yield put(createNewProject());
+  }
+
+  yield call(saveProjectWithKey, state, projectKey);
+}
+
+export function* unArchiveProject({payload: {projectKey}}) {
+  const state = yield select();
+  yield put(changeProject(projectKey));
```

Comment:
This is an interesting case that pushes us right to the limits of the “actions should say what happened” rule. A couple of corollaries of that rule are:

* We should never dispatch two actions from the same execution point
* A saga should never dispatch (`put`) an action merely because it consumed a different action

The latter corollary is violated here.

In the context of my proposal to dispatch `changeCurrentProject` rather than `unArchiveProject` (and probably to remove the `unArchiveProject` action/button altogether), the “what happened” would be simply “the user clicked a project with the intent to switch to it”. Then the system would react to it by:

* Changing the current project (`currentProject` reducer)
* Un-archiving the project (`projects` reducer)
* Persisting the project (`projects` saga, here)

What complicates this is that we really only need to persist the project if we’re switching to a project that was previously archived. However, the way that `redux-saga` works is that sagas’ `take` effects always consume actions _after_ the reducers have done so. So, there’s no straightforward way to know whether the project had previously been archived by the time this code is running.

I can think of a few potential ways around the issue:

1. Create a new action along the lines of `CHANGE_AND_UNARCHIVE_CURRENT_PROJECT`, which the system handles appropriately (by doing all the things it would do for both `CHANGE_CURRENT_PROJECT` and `UNARCHIVE_PROJECT`)
2. Pass extra information in the payload of `CHANGE_CURRENT_PROJECT` to indicate that the project had been archived
3. Always persist the current project when switching to it (unless it is pristine)

My strong preference is for the third option, which I don’t believe has any downsides from a behavior standpoint—at worst we’ll simply be saving data to Firebase that’s identical to the data that’s already been saved.

That said, it is something of a lucky coincidence that we can be insensitive to the previous state of the store when deciding how the saga should behave; it’s easy to imagine a situation in which we really would need to vary behavior based on the previous state. Probably a cross-that-bridge-when-we-get-there situation, but I’m interested to hear your thoughts!